File: src/main/java/thedarkcolour/futuremc/client/gui/GuiLoom.java
package thedarkcolour.futuremc.client.gui;

import net.minecraft.client.gui.inventory.GuiContainer;
import net.minecraft.client.resources.I18n;
import net.minecraft.util.ResourceLocation;
import thedarkcolour.futuremc.FutureMC;
import thedarkcolour.futuremc.container.ContainerLoom;

// TODO
public class GuiLoom extends GuiContainer {
    private static final ResourceLocation BACKGROUND = new ResourceLocation(FutureMC.ID,"textures/gui/loom.png");
    private final ContainerLoom container;

    public GuiLoom(ContainerLoom container) {
        super(container);
        this.container = container;
        container.setInventoryUpdateListener(this::onInventoryUpdate);
    }

    private void onInventoryUpdate() {

    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        super.drawScreen(mouseX, mouseY, partialTicks);
        renderHoveredToolTip(mouseX, mouseY);
    }

    @Override
    protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
        String s = I18n.format("container.Loom");
        fontRenderer.drawString(s, 8, 4, 4210752);
        fontRenderer.drawString(container.getPlayerInv().getDisplayName().getFormattedText(), 8, ySize - 94, 4210752);
    }

    @Override
    protected void drawGuiContainerBackgroundLayer(float partialTicks, int mouseX, int mouseY) {

    }
}
/*
public class GuiLoom extends GuiContainer {
    private static final ResourceLocation BACKGROUND = new ResourceLocation(FutureMC.ID,"textures/gui/loom.png");
    private static final int NUM_BANNER_RECIPES = (BannerPattern.values().length - 5 - 1 + 4 - 1) / 4;
    private static final List<EnumDyeColor> color = Lists.newArrayList(EnumDyeColor.GRAY, EnumDyeColor.WHITE);
    private static final ResourceLocation[] designs = Arrays.stream(BannerPattern.values())
            .filter(pattern -> pattern != BannerPattern.BASE)
            .map(pattern -> BannerTextures.BANNER_DESIGNS.getResourceLocation("b8" + pattern.getHashname() + "15", Lists.newArrayList(BannerPattern.BASE, pattern), color))
            .toArray(ResourceLocation[]::new);
    private final ContainerLoom container;
    private ResourceLocation bannerLocation;
    private boolean canScroll;
    private boolean field_214124_v;
    private boolean isBlank;
    private boolean clickedOnScroll;
    private ItemStack bannerStack = ItemStack.EMPTY;
    private ItemStack colorStack = ItemStack.EMPTY;
    private ItemStack patternStack = ItemStack.EMPTY;
    private int recipeIndexOffset = 1;
    private float sliderProgress;

    public GuiLoom(ContainerLoom container) {
        super(container);
        this.container = container;
        container.setInventoryUpdateListener(this::onInventoryUpdate);
    }

    @Override
    public void updateScreen() {
        super.updateScreen();
    }

    @Override
    public void drawScreen(int mouseX, int mouseY, float partialTicks) {
        super.drawScreen(mouseX, mouseY, partialTicks);
        this.renderHoveredToolTip(mouseX, mouseY);
    }

    @Override
    protected void drawGuiContainerForegroundLayer(int mouseX, int mouseY) {
        String s = I18n.format("container.Loom");
        this.fontRenderer.drawString(s, 8, 4, 4210752);
        this.fontRenderer.drawString(container.getPlayerInv().getDisplayName().getFormattedText(), 8, this.ySize - 96 + 2, 4210752);
    }

    @Override
    protected void drawGuiContainerBackgroundLayer(float partialTicks, int mouseX, int mouseY) {
        drawDefaultBackground();
        GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
        mc.getTextureManager().bindTexture(BACKGROUND);
        int i = (width - xSize) / 2;
        int j = (height - ySize) / 2;
        drawTexturedModalRect(i, j, 0, 0, xSize, ySize);

        Slot output = container.getLoomSlot(3);
        drawSlotBackgrounds(i, j);
        int scrollY = (int)(41.0F * sliderProgress);
        drawTexturedModalRect(i + 119, j + 13 + scrollY, 232 + (canScroll ? 0 : 12), 0, 12, 15);

        // Code for the output preview
        if (bannerLocation != null && !isBlank) {
            GlStateManager.pushMatrix();
            GlStateManager.translate(i + 139, j + 52, 0);
            //GlStateManager.scale(24.0f, -24.0f, 1.0f);
            mc.getTextureManager().bindTexture(bannerLocation);
            zLevel = 20;
            drawTexturedModalRect(i + 141, j + 8, 20, 40, 20, 40);
            zLevel = 0;
            GlStateManager.popMatrix();
        } else if (!isBlank) {
            drawTexturedModalRect(i + output.xPos - 2, j + output.yPos - 2, xSize, 17, 17, 16);
        }

        if (canScroll) {
            int l = i + 60;
            int i1 = j + 13;
            int j1 = recipeIndexOffset + 16;

            for (int k1 = recipeIndexOffset; k1 < j1 && k1 < designs.length - 5; ++k1) {
                int l1 = k1 - recipeIndexOffset;
                int i2 = l + l1 % 4 * 14;
                int j2 = i1 + l1 / 4 * 14;
                mc.getTextureManager().bindTexture(BACKGROUND);
                int k2 = ySize;
                if (k1 == container.getSelectedIndex()) {
                    k2 += 14;
                } else if (mouseX >= i2 && mouseY >= j2 && mouseX < i2 + 14 && mouseY < j2 + 14) {
                    k2 += 28;
                }

                //GlStateManager.pushMatrix();
                //GlStateManager.scale(0.85f, 0.85f, 1.0f);

                drawTexturedModalRect(i2, j2, 0, k2, 14, 14);
                if (designs[k1] != null) {
                    mc.getTextureManager().bindTexture(designs[k1]);
                    drawTexturedModalRect(i2 + 4, j2 + 2, 5, 10, 20, 40);
                }

                //GlStateManager.popMatrix();
            }
        } else if (field_214124_v) {
            int l2 = i + 60;
            int i3 = j + 13;
            mc.getTextureManager().bindTexture(BACKGROUND);
            drawTexturedModalRect(l2, i3, 0, ySize, 14, 14);
            int j3 = container.getSelectedIndex();
            if (designs[j3] != null) {
                mc.getTextureManager().bindTexture(designs[j3]);
                drawTexturedModalRect(l2 + 4, i3 + 2, 5, 10, 20, 40);
            }
        }
    }

    public void drawSlotBackgrounds(int i, int j) {
        Slot banner = container.getLoomSlot(0);
        Slot color = container.getLoomSlot(1);
        Slot pattern = container.getLoomSlot(2);

        if (!banner.getHasStack()) {
            drawTexturedModalRect(i + banner.xPos, j + banner.yPos, xSize, 0, 16, 16);
        }

        if (!color.getHasStack()) {
            drawTexturedModalRect(i + color.xPos, j + color.yPos, xSize + 16, 0, 16, 16);
        }

        if (!pattern.getHasStack()) {
            drawTexturedModalRect(i + pattern.xPos, j + pattern.yPos, xSize + 32, 0, 16, 16);
        }
    }

    @Override
    protected void mouseClicked(int mouseX, int mouseY, int mouseButton) throws IOException {
        clickedOnScroll = false;
        if (canScroll) {
            int i = guiLeft + 60;
            int j = guiTop + 13;
            int k = recipeIndexOffset + 16;

            for (int l = recipeIndexOffset; l < k; ++l) {
                int i1 = l - recipeIndexOffset;
                double x = mouseX - (double)(i + i1 % 4 * 14);
                double y = mouseY - (double)(j + i1 / 4 * 14);
                if (x >= 0.0D && y >= 0.0D && x < 14.0D && y < 14.0D && container.enchantItem(mc.player, l)) {
                    Minecraft.getMinecraft().getSoundHandler().playSound(PositionedSoundRecord.getMasterRecord(SoundEvents.UI_BUTTON_CLICK, 1.0F));
                    mc.playerController.sendEnchantPacket(container.windowId, l);
                    return;
                }
            }

            i = this.guiLeft + 119;
            j = this.guiTop + 9;
            if (mouseX >= (double)i && mouseX < (double)(i + 12) && mouseY >= (double)j && mouseY < (double)(j + 56)) {
                clickedOnScroll = true;
            }
        }
        super.mouseClicked(mouseX, mouseY, mouseButton);
    }

    @Override
    protected void mouseClickMove(int mouseX, int mouseY, int clickedMouseButton, long timeSinceLastClick) {
        if (clickedOnScroll && canScroll) {
            int i = guiTop + 13;
            int j = i + 56;
            sliderProgress = (mouseY - i - 7.5F) / (j - i - 15.0F);
            sliderProgress = MathHelper.clamp(sliderProgress, 0.0F, 1.0F);
            recipeIndexOffset = (int)((sliderProgress * getHiddenRows()) + 0.5D) << 2;
        }
        super.mouseClickMove(mouseX, mouseY, clickedMouseButton, timeSinceLastClick);
    }

    @Override
    public void handleMouseInput() throws IOException {
        super.handleMouseInput();
        int i = Mouse.getEventDWheel();

        if (i != 0 && canScroll) {
            i = (int) Math.signum(i);
            int h = getHiddenRows();
            sliderProgress = (float)((double)sliderProgress - i / (double)h);
            sliderProgress = MathHelper.clamp(sliderProgress, 0.0F, 1.0F);
            recipeIndexOffset = 1 + (int)((double)(sliderProgress * (float)h) + 0.5D) * 4;
        }
    }

    private int getHiddenRows() {
        return NUM_BANNER_RECIPES - 4;
    }

    @Override
    protected boolean hasClickedOutside(int param1, int param2, int param3, int param4) {
        return param1 < param3 || param2 < param4 || param1 >= param3 + xSize || param2 >= param4 + ySize;
    }

    private void onInventoryUpdate() {
        ItemStack itemstack = container.getOutput();
        if (itemstack.isEmpty()) {
            bannerLocation = null;
        } else {
            TileEntityBanner te = new TileEntityBanner();
            te.setItemValues(itemstack, false);
            bannerLocation = BannerTextures.BANNER_DESIGNS.getResourceLocation(te.getPatternResourceLocation(), te.getPatternList(), te.getColorList());
        }

        ItemStack stack = container.getBanner();
        ItemStack stack1 = container.getColor();
        ItemStack stack2 = container.getPattern();
        NBTTagCompound tag = stack.getOrCreateSubCompound("BlockEntityTag");
        isBlank = tag.hasKey("Patterns", 9) && !stack.isEmpty() && canFitPatterns(tag.getTagList("Patterns", 10));
        if (isBlank) {
            bannerLocation = null;
        }

        if (!ItemStack.areItemStacksEqual(stack, bannerStack) || !ItemStack.areItemStacksEqual(stack1, this.colorStack) || !ItemStack.areItemStacksEqual(stack2, this.patternStack)) {
            canScroll = !stack.isEmpty() && !stack1.isEmpty() && stack2.isEmpty() && !isBlank;
            field_214124_v = !isBlank && !stack2.isEmpty() && !stack.isEmpty() && !stack1.isEmpty();
        }

        bannerStack = stack.copy();
        colorStack = stack1.copy();
        patternStack = stack2.copy();
    }

    private boolean canFitPatterns(NBTTagList list) {
        int i = 0;

        for (NBTBase ignored : list) {
            ++i;
        }
        return i <= 6;
    }
}
*/

File: src/main/java/thedarkcolour/futuremc/client/gui/GuiLoom.kt
package thedarkcolour.futuremc.client.gui

import net.minecraft.client.Minecraft
import net.minecraft.client.audio.PositionedSoundRecord
import net.minecraft.client.renderer.BannerTextures
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.client.resources.I18n
import net.minecraft.init.SoundEvents
import net.minecraft.item.EnumDyeColor.GRAY
import net.minecraft.item.EnumDyeColor.WHITE
import net.minecraft.item.ItemStack
import net.minecraft.tileentity.BannerPattern
import net.minecraft.tileentity.TileEntityBanner
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.MathHelper
import org.lwjgl.input.Mouse
import thedarkcolour.futuremc.FutureMC
import thedarkcolour.futuremc.container.ContainerLoom
import kotlin.math.ceil
import kotlin.math.sign

class GuiLoom(container: ContainerLoom) : FGui<ContainerLoom>(container) {
    private var preview: ResourceLocation? = null
    private var banner = ItemStack.EMPTY
    private var color = ItemStack.EMPTY
    private var pattern = ItemStack.EMPTY
    private var isCrafting = false
    private var hasBannerPattern = false
    private var cannotFitPatterns = false
    private var sliderProgress = 0f
    private var clickedOnScroll = false
    private var recipeIndexOffset = 1

    init {
        container.setInventoryUpdateListener { onInventoryUpdate() }
    }

    override fun drawGuiContainerForegroundLayer(mouseX: Int, mouseY: Int) {
        fontRenderer.drawString(I18n.format("container.loom"), 8, 4, 4210752)
        fontRenderer.drawString(getContainer().playerInv.displayName.unformattedText, 8, ySize - 94, 4210752)
    }

    override fun drawGuiContainerBackgroundLayer(partialTicks: Float, mouseX: Int, mouseY: Int) {
        GlStateManager.color(1.0f, 1.0f, 1.0f, 1.0f)
        mc.textureManager.bindTexture(BACKGROUND)
        val i = guiLeft
        val j = guiTop
        drawTexturedModalRect(i, j, 0, 0, xSize, ySize)
        val bannerSlot = getContainer().getLoomSlot(0)
        val dyeSlot = getContainer().getLoomSlot(1)
        val patternSlot = getContainer().getLoomSlot(2)
        val outputSlot = getContainer().getLoomSlot(3)
        if (!bannerSlot.hasStack) {
            drawTexturedModalRect(i + bannerSlot.xPos, j + bannerSlot.yPos, xSize, 0, 16, 16)
        }

        if (!dyeSlot.hasStack) {
            drawTexturedModalRect(i + dyeSlot.xPos, j + dyeSlot.yPos, xSize + 16, 0, 16, 16)
        }

        if (!patternSlot.hasStack) {
            drawTexturedModalRect(i + patternSlot.xPos, j + patternSlot.yPos, xSize + 32, 0, 16, 16)
        }

        val k = (41.0f * sliderProgress).toInt()
        drawTexturedModalRect(i + 119, j + 13 + k, 232 + if (isCrafting) 0 else 12, 0, 12, 15)
        if (preview != null && !cannotFitPatterns) {
            mc.textureManager.bindTexture(preview!!)
            drawScaledCustomSizeModalRect(i + 141, j + 8, 1f, 1f, 20, 40, 20, 40, 64f, 64f)
        } else if (cannotFitPatterns) {
            drawTexturedModalRect(i + outputSlot.xPos - 2, j + outputSlot.yPos - 2, xSize, 17, 17, 16)
        }

        if (bannerSlot.hasStack) {
            val listLeft = i + 60
            val listTop = j + 13

            if (isCrafting) {
                val lastVisibleItem = recipeIndexOffset + 16
                var index = recipeIndexOffset

                while (index < lastVisibleItem && index <= ContainerLoom.BASIC_PATTERNS.size) {
                    val listIndex = index - recipeIndexOffset
                    val slotLeft = listLeft + listIndex % 4 * 14
                    val slotTop = listTop + listIndex / 4 * 14

                    mc.textureManager.bindTexture(BACKGROUND)
                    var k2 = ySize
                    if (index == getContainer().getSelectedIndex()) {
                        k2 += 14
                    } else if (((mouseX - slotLeft) in 0..14) && ((mouseY - slotTop) in 0..14)) {
                        k2 += 28
                    }
                    drawTexturedModalRect(slotLeft, slotTop, 0, k2, 14, 14)

                    if (BASIC_TEXTURES[index - 1] != null) {
                        mc.textureManager.bindTexture(BASIC_TEXTURES[index - 1]!!)
                        drawScaledCustomSizeModalRect(slotLeft + 4, slotTop + 2, 1f, 1f, 20, 40, 5, 10, 64.0f, 64.0f)
                    }
                    ++index
                }
            } else if (hasBannerPattern) {
                val index = getContainer().getSelectedIndex()

                mc.textureManager.bindTexture(BACKGROUND)
                drawTexturedModalRect(listLeft, listTop, 0, ySize, 14, 14)

                if (PATTERN_TEXTURES[index] != null) {
                    mc.textureManager.bindTexture(PATTERN_TEXTURES[index]!!)
                    drawScaledCustomSizeModalRect(listLeft + 4, listTop + 2, 1f, 1f, 20, 40, 5, 10, 64f, 64f)
                }
            }
        }
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        clickedOnScroll = false

        if (isCrafting) {
            var listLeft = guiLeft + 60
            var listTop = guiTop + 13
            val lastVisibleItem = recipeIndexOffset + 16

            for (i in recipeIndexOffset until lastVisibleItem) {
                val listIndex = i - recipeIndexOffset
                val mX = mouseX - (listLeft + listIndex % 4 * 14).toDouble()
                val mY = mouseY - (listTop + listIndex / 4 * 14).toDouble()

                if (mX in 0.0..14.0 && mY in 0.0..14.0 && getContainer().enchantItem(mc.player, i)) {
                    Minecraft.getMinecraft().soundHandler.playSound(PositionedSoundRecord.getMasterRecord(SoundEvents.UI_BUTTON_CLICK, 1.0f))
                    mc.playerController.sendEnchantPacket(getContainer().windowId, i)
                    return
                }
            }
            listLeft = guiLeft + 119
            listTop = guiTop + 9
            if (mouseX >= listLeft.toDouble() && mouseX < (listLeft + 12) && mouseY >= listTop.toDouble() && mouseY < (listTop + 56).toDouble()) {
                clickedOnScroll = true
            }
        }

        super.mouseClicked(mouseX, mouseY, mouseButton)
    }

    override fun mouseClickMove(mouseX: Int, mouseY: Int, clickedMouseButton: Int, timeSinceLastClick: Long) {
        if (clickedOnScroll && isCrafting) {
            val i = guiTop + 13
            val j = i + 56
            sliderProgress = (mouseY - i - 7.5f) / (j - i - 15.0f)
            sliderProgress = MathHelper.clamp(sliderProgress, 0.0f, 1.0f)
            val k = ROWS - 4
            val l = ((sliderProgress * k.toFloat()).toDouble() + 0.5).toInt().coerceAtLeast(0)

            recipeIndexOffset = 1 + (l * 4)
        } else {
            super.mouseClickMove(mouseX, mouseY, clickedMouseButton, timeSinceLastClick)
        }
    }

    override fun handleMouseInput() {
        super.handleMouseInput()
        if (isCrafting) {
            val i = sign(Mouse.getEventDWheel().toDouble())
            val j = ROWS - 4
            sliderProgress = (sliderProgress - i / j).toFloat()
            sliderProgress = MathHelper.clamp(sliderProgress, 0f, 1f)
            recipeIndexOffset = 1 + (((sliderProgress * j) + 0.5).toInt() shl 2)
        }
    }

    override fun hasClickedOutside(mouseX: Int, mouseY: Int, guiLeft: Int, guiTop: Int): Boolean {
        return mouseX < guiLeft || mouseY < guiTop || mouseX >= (guiLeft + xSize) || mouseY >= (guiTop + ySize)
    }

    private fun onInventoryUpdate() {
        val stack = getContainer().output

        preview = if (stack.isEmpty) {
            null
        } else {
            val tile = TileEntityBanner()
            tile.setItemValues(stack, false)
            BannerTextures.BANNER_DESIGNS.getResourceLocation(tile.patternResourceLocation, tile.patternList, tile.colorList)
        }

        val banner = getContainer().banner
        val color = getContainer().color
        val pattern = getContainer().pattern
        val nbt = banner.getOrCreateSubCompound("BlockEntityTag")
        cannotFitPatterns = nbt.hasKey("Pattern", 9) && !banner.isEmpty && nbt.getTagList("Patterns", 10).tagCount() >= 6
        if (cannotFitPatterns) {
            preview = null
        }

        if (!ItemStack.areItemStacksEqual(banner, this.banner) || !ItemStack.areItemStacksEqual(color, this.color) || !ItemStack.areItemStacksEqual(pattern, this.pattern)) {
            isCrafting = !banner.isEmpty && !color.isEmpty && pattern.isEmpty && !cannotFitPatterns
            hasBannerPattern = !banner.isEmpty && !color.isEmpty && !pattern.isEmpty && !cannotFitPatterns
        }

        this.banner = banner.copy()
        this.color = color.copy()
        this.pattern = pattern.copy()
    }

    companion object {
        private val BACKGROUND = ResourceLocation(FutureMC.ID, "textures/gui/loom.png")
        private val ROWS = ceil(ContainerLoom.BASIC_PATTERNS.size / 4.0f).toInt()
        // Used for previews in the list
        private val PALETTE = listOf(GRAY, WHITE)

        // Contains all pattern textures
        private val PATTERN_TEXTURES = arrayListOf<ResourceLocation?>()
        // Contains only regular textures
        private val BASIC_TEXTURES = arrayListOf<ResourceLocation?>()

        init {
            for (pattern in BannerPattern.values()) {
                val s = "b" + GRAY.metadata
                val s1 = pattern.hashname + WHITE.metadata

                val texture = BannerTextures.BANNER_DESIGNS.getResourceLocation(s + s1, listOf(BannerPattern.BASE, pattern), PALETTE)

                if (texture == null) {
                    FutureMC.LOGGER.error("Failed to load texture for banner pattern: $pattern")
                }

                PATTERN_TEXTURES.add(texture)
                if (ContainerLoom.BASIC_PATTERNS.contains(pattern)) {
                    BASIC_TEXTURES.add(texture)
                }
            }
        }
    }
}